### PR TITLE
Improve the print method

### LINKED ISSUE #290: Consider an improved print method
* From #266: not informative for zero-length outputs
* From #283: hard to tell which element is which

Could fix using logic from `stringr::str_view()` plus better behaviour for 0-length output.

### PR DESCRIPTION
**Pull request by hadley**
Inspired by stringr::str_view()

Fixes #290

### CODE REVIEW

File: R/glue.R [hunk 1]
```diff
@@ -287,12 +287,21 @@ trim <- function(x) {
 
 #' @export
 print.glue <- function(x, ..., sep = "\n") {
-  x[is.na(x)] <- style_na(x[is.na(x)])
-
-  if (length(x) > 0) {
-    cat(x, ..., sep = sep)
+  if (length(x) == 0) {
+    cat("<Length-0 glue vector>\n")
```
Comment (DavisVaughan) [inline R/glue.R:291]: Screen Shot 2023-01-27 at 4 03 14 PM

You could use the same idea as vctrs and always show `<glue[n]>` so the 0 length case feels more natural

File: R/glue.R [hunk 2]
```diff
@@ -287,12 +287,21 @@ trim <- function(x) {
...
+  id <- format(paste0("[", seq_along(x), "] "), justify = "right")
+  indent <- paste0(style_grey(id, "|"), " ")
+  exdent <- paste0(strrep(" ", nchar(id[[1]])), style_grey("|"), " ")
+
+  out <- x
+  out[is.na(out)] <- style_red("NA")
+  out <- paste0(indent, out)
+  out <- gsub("\n", paste0("\n", exdent), out)
+  cat(paste0(out, "\n", collapse = ""))
+
```
Comment (DavisVaughan) [inline R/glue.R:304]: I think we should consider a max.print default like base R print methods have.

If you accidentally print `as_glue(1:1e6)` it takes around 10 seconds to print out for me.

But that will probably be finicky enough that it should be in a separate PR

File: R/glue.R [hunk 3]
```diff
@@ -287,12 +287,21 @@ trim <- function(x) {
...
   }
 
+  id <- format(paste0("[", seq_along(x), "] "), justify = "right")
+  indent <- paste0(style_grey(id, "|"), " ")
+  exdent <- paste0(strrep(" ", nchar(id[[1]])), style_grey("|"), " ")
+
+  out <- x
+  out[is.na(out)] <- style_red("NA")
+  out <- paste0(indent, out)
+  out <- gsub("\n", paste0("\n", exdent), out)
```
Comment (DavisVaughan) [inline R/glue.R:302]: I tried a number of things to break this. The only thing I can think of is that it can make colored values show up incorrectly

```r
# cli::col_red("foo\nbar")
x <- "\033[31mfoo\nbar\033[39m"
cat(x)

glue::glue(x)
```
Screen Shot 2023-01-27 at 4 16 55 PM

**Review by DavisVaughan — APPROVED** [review]

**Comment by jennybc** [comment]
This feels like a pretty big change in the print method to me! So much so that it gives me pause.

I've always assumed that the main benefit of having the `"glue"` class was to get the quote-free printing and this walks away from that. I know we have some regrets about the `"glue"` class, but this seems to undermine its main rationale?

This almost feels more like a `str()` method. What do you think of me turning it into that?

(And, if so, yes I still think we need to do something when printing a glue object of length 0.)

File: R/glue.R (hunk 3 above)
Comment (jennybc) [reply R/glue.R:302]: Presumably we can fix this if we set our minds to it, although it would be much more pleasant to do so using cli, which is only in `Suggests`.

**Review by jennybc — COMMENTED** [review]
This feels like a pretty big change in the print method to me! So much so that it gives me pause.

I've always assumed that the main benefit of having the "glue" class was to get the quote-free, clutter-free printing and this walks away from that. I know we have some regrets about the "glue" class, but this seems to undermine its main rationale?

This almost feels more like a str() method. What do you think of me turning it into that?

(And, if so, yes I still think we need to do something when printing a glue object of length 0.)

**Comment by jennybc** [comment]
I ran revdep checks and, as expected, only a saw a couple of failures due to this, including https://github.com/r-lib/testthat/issues/1868. But in the course of figuring out that puzzle, I checked pillar locally and saw _massive_ snapshot test change, which I think could be pretty widespread. I know that won't make tests fail on CRAN, in general, but it increases my hunch that we should introduce this display as a `str()` method.

**Comment by hadley** [comment]
We should maybe talk this through live as it's definitely not a `str` method. I think it stays true to the intent of glue to just "show the strings", while avoiding confusion in cases like this:

``` r
glue::glue("{x}", x = c("a", "b", "c\nd"))
#> a
#> b
#> c
#> d
```

<sup>Created on 2023-09-25 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

**Comment by hadley** [comment]
I've resolved the remaining discussion items; @jennybc we should chat to confirm that this makes sense.

File: R/glue.R [hunk 4]
```diff
@@ -289,11 +289,35 @@ trim <- function(x) {
...
-  x[is.na(x)] <- style_na(x[is.na(x)])
+print.glue <- function(x, ..., sep = "\n", max = getOption("max.print")) {
+  cat("<glue[", length(x), "]>\n", sep = "")
+  if (length(x) == 0) {
+    return(invisible())
+  }
+
+  truncated <- length(x) > max
+  if (truncated) {
+    x <- x[seq_len(max)]
```
Comment (jennybc) [inline R/glue.R:300]: Do we want this to affect the return value? I.e. `print()` no longer passes `x` through invisibly.

**Review by jennybc — APPROVED** [review]
I'm still not sure I'm on board with this, but we'll discuss in person. I'm comfortable with how it's done.

**Comment by hadley** [comment]
See discussion in #290
